### PR TITLE
CircleButton and UserAvatar components

### PR DESCRIPTION
**Pull request by lemald**
Asana ticket: [⚙️ Create component for email address initial icon](https://app.asana.com/0/1148853526253420/1206526457565579/f)

### CODE REVIEW

File: assets/src/components/circleButton.tsx
Comment (joshlarson) [inline assets/src/components/circleButton.tsx:14]: Should this take arguments to specify its size? Either `small`/`medium`/`large` or `height`/`width`? Or is there a way to do that that I missed?

Comment (lemald) [reply assets/src/components/circleButton.tsx:14]: From the Figma, there are numerous sizes of the user avatar, but only the one size of the circle button. After discussion with design I decided to make the avatar fill its container and make the button a fixed size.

File: assets/stories/skate-components/userAvatar.stories.tsx
Comment (joshlarson) [inline assets/stories/skate-components/userAvatar.stories.tsx:20]: Nit: It might be helpful to nest these (along with the upcoming `<LoggedInAs />` component) into a subfolder, rather than having them all live top-level. I called it `identity`, but I'm not overly attached to that name.


Screenshot 2024-02-12 at 1 48 51 PM

Just thinking that the top level of ☝️ will get a bit cluttered.

Comment (lemald) [reply assets/stories/skate-components/userAvatar.stories.tsx:20]: I'd be open to moving UserAvatar, though CircleButton isn't necessarily tied to the user avatar (and I think design envisions using it in other places in the future).

File: assets/stories/skate-components/userAvatar.stories.tsx
Comment (lemald) [reply assets/stories/skate-components/userAvatar.stories.tsx:7]: I'm sort of confused - you should be able to just type in an email in the UserAvatar Storybook entry. Are you thinking of CircleIcon? I don't really want to put stuff about the inner workings of the UserAvatar component in the stories for CircleIcon.

Comment (joshlarson) [reply assets/stories/skate-components/userAvatar.stories.tsx:7]: Yes, I'm sorry - I mixed up which was which between `CircleButton` and `UserAvatar`, and put my comments in the wrong place. 

I'll migrate over to `circleAvatar.stories.tsx` to continue this :)

File: assets/stories/skate-components/circleButton.stories.tsx
Comment (joshlarson) [inline assets/stories/skate-components/circleButton.stories.tsx:15]: I originally wrote my comment (on the wrong story) when the `children` component was exposed as an argument, which made the arguments table kind of a mess. I don't feel as strongly about it right now, but there are two things about the UX here that kind of throw me off, and I imagine that they'd throw a non-technical user off as well.

1. The default story is something you'd never actually see in production. Right now, the only thing we'll ever see live is a `CircleButton` with a `UserAvatar` inside, and even in the future, we'll always put _something_ inside.
2. There's no way to change the letter rendered inside the icon on this page. Not the biggest deal on this page, because of the `UserAvatar` stories, but I also don't think it's the worst thing in the world to allow this to be more customizable.

Comment (joshlarson) [reply assets/stories/skate-components/circleButton.stories.tsx:15]: To take my two piddly comments above, and cast them in a wildly overblown way, both of ☝️ points reflect the question of whether Storybook is a tool for dev, or a tool for people other than dev.

If we place boundaries around what types of things we allow in our Storybook that really only make sense if you know the inner workings of those icons (for instance, the fact that the user icon thing is implemented using a `children` attribute), then that places Storybook as a dev-only tool. 

Another way of putting it is that if an object has any sort of nontrivial interface, then either the story code needs to be hyper-aware of that nontrivial interface (by using a decorator), or the person clicking around in Storybook needs to be aware (by seeing an interface that either lacks options or displays things about JSXNodes). I would prefer to have the Story code itself be the thing that's aware, and let people interacting with Storybook have a better experience.

Comment (lemald) [reply assets/stories/skate-components/circleButton.stories.tsx:15]: So for what it's worth I was originally conceiving of this all as one component but after talking to design I realized that they actually saw it as two different components, so I think in this case it does make sense to break it apart like this (though I agree in general that UX of Storybook for non-devs should at least be a consideration).

Comment (joshlarson) [reply assets/stories/skate-components/circleButton.stories.tsx:15]: Huh, fair enough.

Comment (firestack) [reply assets/stories/skate-components/circleButton.stories.tsx:15]: Maybe we should start organizing stories according to [atomic design](https://bradfrost.com/blog/post/atomic-web-design/)?
This seems like it'd be a molecule?

Comment (firestack) [reply assets/stories/skate-components/circleButton.stories.tsx:15]: It'd maybe make it more clear on how granular a component is if it were organized according to atomic design, which I see as addressing point 1

> 1. The default story is something you'd never actually see in production. Right now, the only thing we'll ever see live is a `CircleButton` with a `UserAvatar` inside, and even in the future, we'll always put _something_ inside.

File: assets/css/_circle_button.scss
Comment (joshlarson) [inline assets/css/_circle_button.scss:13]: I could be wrong, but I think this is applying the focus border when you click on the button, which ... I don't _think_ that's what focus state is supposed to do? I think the behavior is supposed to be that we get only the "active" state (the darker circle around the avatar) on click, not the focus outline as well.

Comment (lemald) [reply assets/css/_circle_button.scss:13]: The thing is that the active state is actually orthogonal from the hover / focus states, and it just so happens that clicking on the button both focuses it and toggles the active state. I've asked design how they want the precedence to work when we're in both states, but for now I'm letting the interaction states override the active / inactive state since the other way around seems like it would be worse for usability.

File: assets/css/_circle_button.scss
Comment (lemald) [reply assets/css/_circle_button.scss:21]: That is a good point about the grays, I'll bring this up when I talk with design about the purples as well.

File: assets/stories/skate-components/circleButton.stories.tsx
Comment (lemald) [reply assets/stories/skate-components/circleButton.stories.tsx:15]: If we feel that the solution is to pursue a large reorganization of our Storybook stories according to atomic design, is that something we can look into in a followup task?

Comment (firestack) [reply assets/stories/skate-components/circleButton.stories.tsx:15]: @lemald https://app.asana.com/0/1148853526253420/1206580891011553/f

Comment (lemald) [reply assets/stories/skate-components/circleButton.stories.tsx:15]: Thanks!

**Comment by joshlarson** [comment]
I'm not approving yet because of the two design-related unresolved questions around focus state and the background gray-200 color of the active state. The rest looks good to me.

I do think that it would be good for us to get a clearer sense as a team of what our storybook stories should look like, but that discussion doesn't need to block this PR. Plus, design should probably be involved in that anyway, since our goal is for this to be a tool at least partly for them.

File: assets/css/_circle_button.scss
Comment (lemald) [reply assets/css/_circle_button.scss]: See 850d2029 (changed to use the new grays).

File: assets/css/_circle_button.scss
Comment (lemald) [reply assets/css/_circle_button.scss]: Changed to use `:focus-visible` so the focus state only appears when tabbing.

File: assets/css/_circle_button.scss
Comment (joshlarson) [inline assets/css/_circle_button.scss:11]: One additional nit - it looks like the hover state is overriding the active state, which I doubt is what design intended (it feels pretty weird to me). I can't quite wrap my mind around _why_ it's happening, since the nav menu looks like its CSS is set up pretty similarly, but it's not exhibiting this behavior.

That said, I don't think we should block merging for more than a very short time on this, so if we can get design feedback quickly on it, and/or if there's an obvious fix, then hurrah, but otherwise I'm good to merge and follow up with a fix (if indeed a fix is required)

File: assets/css/_circle_button.scss
Comment (firestack) [inline assets/css/_circle_button.scss:15]: nit(non-blocking): just wondering if the border utilities would've worked here? https://getbootstrap.com/docs/5.3/utilities/borders/ seems like it would've been classes `border` and `border-primary`?

(comment originally authored at https://github.com/mbta/skate/commit/850d20296ea5d5dcc22375afcf1a7277c458f58a#r138593057 )

Comment (firestack) [reply assets/css/_circle_button.scss:15]: Though I guess this would've had to have been 
```suggestion
    @extend .border,.border-primary;
```
because of the `:focus-visible` state, unless bootstrap has like, "is focus visible" utilities that generate a `border-focus-visible` and `border-primary-focus-visible`

Comment (lemald) [reply assets/css/_circle_button.scss:15]: Yeah... maybe possible (I'd like to make it work) but seems like it might be too much effort for the moment.

File: assets/css/_circle_button.scss
Comment (joshlarson) [reply assets/css/_circle_button.scss:11]: Ooooh I think I see the problem... sort of. It has to do with how CSS decides how to interpret conflicting rules.

File: assets/css/_circle_button.scss
Comment (joshlarson) [inline assets/css/_circle_button.scss]:
```suggestion
  &.c-circle-button--active {
```

When there are conflicting rules, like `:hover` versus a specific class, CSS will pick whichever rule trigger is more specific. So without applying this, the rule trigger was `.c-circle-button:hover` (class plus hover state) versus `.c-circle-button--active` (just a class). With applying this change, now it would be `.c-circle-button .c-circle-button--active` (two classes), which I guess CSS sees as more specific than `.c-circle-button:hover`

Comment (joshlarson) [reply assets/css/_circle_button.scss]: lol great minds

**Review by joshlarson — APPROVED** [review]

**Pull request merged**
